src/vset.c

```diff
@@ -0,0 +1,2094 @@
+#include "vset.h"
+#include "rax.h"
+#include "endianconv.h"
+#include "serverassert.h"
+#include "hashtable.h"
+#include "util.h"
+#include "zmalloc.h"
+
+#include <string.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+/*************************************************************************************************************
+ *                                pVector Implementation
+ *************************************************************************************************************/
+
+#define PV_CARD_BITS 30
+#define PV_ALLOC_BITS 34
+#define PV_MAX_ELEMENTS ((1ULL << PV_CARD_BITS) - 1)
+#define PV_HEADER_SIZE (sizeof(pVector))
+#define PV_ELEM_SIZE (sizeof(void *))
+#define PV_ALLOC(pv) (pv ? pv->alloc : 0)
+#define PV_LEN(pv) (pv ? pv->len : 0)
+#define PV_USED_SIZE(pv) (pv ? (PV_HEADER_SIZE + (pvLen(pv)) * PV_ELEM_SIZE) : 0)
+
+/* Custom vector structure with embedded allocation and length counters */
+typedef struct {
+    uint64_t len : 30;   /* Number of elements (cardinality) */
+    uint64_t alloc : 34; /* Allocated memory (zmalloc_size of the current vector allocation) */
+    void *data[];        /* Flexible array member */
+} pVector;
+
+/* Returns the number of elements currently stored in the pVector.
+ *
+ * Arguments:
+ *   vec - The pVector to query.
+ *
+ * Return:
+ *   The number of elements in the vector.
+ *   Note that a NULL is a !!!valid!!! vector - returns 0 if the vector is NULL. */
+static inline uint32_t pvLen(pVector *vec) {
+    return PV_LEN(vec);
+}
+
+/* Ensures that a pVector has enough capacity to hold additional elements.
+ *
+ * This function guarantees that the given pVector `pv` has at least enough
+ * allocated space to accommodate `additional` more elements, growing it if necessary.
+ * If the vector is currently `NULL`, it will be newly allocated.
+ *
+ * The allocation is handled using `zmalloc` or `zrealloc_usable`, depending on whether
+ * the vector is new or already initialized. The internal `alloc` field is updated to
+ * reflect the actual allocated size.
+ *
+ * Arguments:
+ *   pv       - Pointer to an existing pVector or NULL.
+ *   additional - The number of additional elements the vector should be able to accommodate.
+ *
+ * Return:
+ *   A pointer to the resized (or newly allocated) pVector with sufficient capacity.
+ *   Returns NULL only if the allocation fails.
+ *
+ * Note:
+ *   The `additional` is the number of *additional* elements beyond the current length.
+ *   This function does not modify the vector's logical length (`len`), only its allocation. */
+pVector *pvMakeRoomFor(pVector *pv, size_t additional) {
+    if (additional == 0) return pv;
```

Comment:
There is no reasonable bounds checking on `additional`.  We currently have 30 bits for `len`, and there's no check for overflow here.  Even so, it's unrealistic to think that this data structure would/should accommodate 1B entries - the memory moves would be problematic long before reaching that.

Can we add an assertion that the length isn't being extended too large - which would indicate some type of programming error.  What's a reasonable max for this data structure?  10K entries?  1M entries?  10M?

Reply:
yes. will add them